Feat/enhanced change listener

davidgarciaanton opened the pull request:
As EncryptionProtocol is not accessible and there may be the requirement to respond only to specific keys changed, and the regular `SharedPreferences.OnPreferenceChangeListener` callback isn't enough as we get the stored version of the key, witch we cannot compare against constants.

I've created a specific `OnSecurePreferenceChangeListener` that gives a facility  to at least compare the changed key against some arbitrary string.

I don't know in terms of security of implementation if this is the best way to achieve this. Please let me know if you think of another way more secure

patrickfav commented:
Thanks for the PR. I will review it in the next days!

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/SecureSharedPreferences.java:
```diff
@@ -253,6 +256,45 @@ public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeL
         sharedPreferences.unregisterOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener);
     }
 
+    private final List<SharedPreferenceChangeListenerWrapper> securePreferenceListeners = new LinkedList<>();
```
Why do we need an additional list managing the `securePreferenceListeners`? Wouldn't it suffice to just wrap it in `new SharedPreferenceChangeListenerWrapper();` and add it to the internal listener map? This would remove a lot of the complexity of this code.

davidgarciaanton replied:
What I wanted to achieve is to allow the client code to register/unregister using the original `OnSecurePreferenceChangeListener` and if I just wrapped this one with the `SharedPreferenceChangeListenerWrapper` and pass it down to the _SharedPreference's own listener management_ I didn't know how to do the **unregister** part. Do you have an idea for doing it? I think I might be missing something here.

patrickfav replied:
Fair enough, wrapping it would make it another reference the user does not know, so the unregister wouldn't work. I'd prefer the cleaner shared preference managed version, but do not have a better suggestions.

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/SecureSharedPreferences.java:
```diff
@@ -253,6 +256,45 @@ public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeL
+                    unregisterOnSharedPreferenceChangeListener(listenerWrapper);
+                    iterator.remove();
+                } else if (wrapped == listener) {
+                    found = true; // we let the loop continue, so that we may cleanup some more references
```
Why add the complexity of checking if the listener reference is already added? The default implementation also doesn't care about this. I get stable behaviour, but adding the same reference twice would be a naiive developer error. See: https://android.googlesource.com/platform/frameworks/base.git/+/master/core/java/android/app/SharedPreferencesImpl.java

davidgarciaanton replied:
Here as I was also doing some cleanup for released references I though that it wouldn't do harm to also add the safety check.

I can remove it, and just asume developer would do the correct thing. Do you recommend to also remove weak references cleanup loop?

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/SharedPreferenceChangeListenerWrapper.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private final OnSecurePreferenceChangeListener.DerivedKeyComparison keyComparison;
+    private final WeakReference<OnSecurePreferenceChangeListener> wrappedRef;
```
I think using `WeakReference `here might lead to problems. This may be not the behaviour you want and it is quite difficult to understand from the outside why this may not work for a developer. I usually refrain from using Weak/Soft references as they not very clear to most devs how they work (and behave differently on Android ie. more aggresivly). I would make it the callers responsibility to unregister his listeners, as it is usually the case in Android.
As reference why this might be an issue: https://android.jlelse.eu/android-weakreferences-how-not-to-5522bb325935

davidgarciaanton replied:
I know and I myself did suffer some issues at this respect letting Kotlin's lambda to implement directly a `OnSharedPreferencesChangeListener` which was translated to an anonymous class instantiation, thus being garbage collected (I didn't know at the time how kotlin translated that piece of code, I learnt it by force).

I decided to go with the `WeakReference` approach just based on the same principle than the `SharedPreferences`, because well the system indeed treats `SharedPreferences` instances as singletons, then adding listeners and failing to unregister could cause potential leak. 

For me is a design choice, if you prefer I can remove the WeakReferences, but here I like them indeed, for same reason than a user should know well to not register twice a listener, they should know well to keep a reference to the listener (I could add this fact to the javadoc). It's also true that armadillo isn't enforcing any singleton and will be the user of the library who decides how he/she manages that.

patrickfav replied:
Fair argument, however I prefer the easier to understand, explicit way with a little more boilerplate, ie. please change it to register/unregister with a strong reference.

davidgarciaanton replied:
Ok, I've just updated to avoid usage of WeakReferences on listeners

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java:
```diff
@@ -0,0 +1,34 @@
+     * @param sharedPreferences the shared preferences instance
+     * @param derivedKeyChanged the value of the changed key in it's derived form (as it's actually stored)
+     */
+    void onSecurePreferenceChanged(@NonNull DerivedKeyComparison comparison, @NonNull SharedPreferences sharedPreferences, @NonNull String derivedKeyChanged);
```
`@NonNull String derivedKeyChanged` >  I would call it `@NonNull String derivedContentKey`

patrickfav replied:
Keeping it same with the original `OnSharedPreferenceChangeListener`, I would order the parameters

`(@NonNull SharedPreferences sharedPreferences,  @NonNull String derivedKeyChanged, @NonNull DerivedKeyComparison comparison)`

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * Allows client side {@link OnSecurePreferenceChangeListener} to check if the changed key is the key of interest.
+     */
+    interface DerivedKeyComparison {
```
Instead of providing the caller with a generic `doesMyKeyMatchWithTheDerivedKey()` it may be easier to understand for the callback to provide a checker which has the derived key backed in. Therefore the onSecurePreferenceChanged()` would not require to pass the `String derivedKeyChanged` argument and the caller would just check with e.g. `comparison.matchesDerivedKey(String rawContentKey)`

davidgarciaanton replied:
Yes! Here's the funny story...

I begun doing it that way indeed, but then I though that in case of positive match I could use the `sharedPreferences` and `key` parameters, to get the changed value.

I then realised that this didn't work because the `sharedPreferences` object was the original one, and requesting a preference with the `key` parameter would provide the _encrypted_ value.

And then I probably left that as it was. I think that it might be nice to have both of the proposals. So let's see what do you think about having the method signature 

`void onSecurePreferenceChanged(@NonNull SharedPreferences sharedPreferences, @NonNull DerivedKeyComparison comparison)`

Where the **instance** of sharedPreferences would be the `SecureSharedPreferences` instead of the underlying `SharedPreferencesImpl` so that client code could then get the changed value.

I didn't do that at the time, worried about security or something similar. If we're not interesting on allowing such behaviour, then I may suggest to change the signature of the method to just receive the `DerivedKeyComparison` parameter as right now is the only usable parameter

patrickfav replied:
I like the improved version. What we should change is that the `SharedPreferences sharedPreferences` should be the Armadillo instance not the inner wrapped version. A) a lib user has no business changing the raw encrypted data B) the raw encrypted shared pref is practically useless for the user, with the Armadillo wrapper she would be able to retrieve the enecrypted data with the content-keys she knows.

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java:
```diff
@@ -0,0 +1,34 @@
+public interface OnSecurePreferenceChangeListener {
+
+    /**
+     * Variation of the regular {@link android.content.SharedPreferences.OnSharedPreferenceChangeListener#onSharedPreferenceChanged(SharedPreferences, String)} that
```
Please add a simple code example in javadoc on how to use it.

`<pre> .... </pre>`

davidgarciaanton commented on armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * Allow to register preference change listeners that receive
+ * @since 03.07.2019
```
Ok, for this one I just applied some of the existing code patterns which used dates. Do you prefer if I remove the @since tag, as I don't actually know next release version. Or either you tell me if it's OK to say it will be `0.9.0`?

patrickfav replied:
You can set 0.9.0 or just remove it, whichever you prefer.

davidgarciaanton commented on armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java:
```diff
@@ -0,0 +1,34 @@
+import android.support.annotation.NonNull;
+
+/**
+ * Allow to register preference change listeners that receive
```
Probably, I will give it a look

davidgarciaanton commented:
I'll try to give it another round when't I find a moment to it. Thanks for taking your time to review!

davidgarciaanton commented on armadillo/src/main/java/at/favre/lib/armadillo/SharedPreferenceChangeListenerWrapper.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * Implementation that will wrap a {@link OnSecurePreferenceChangeListener} to adapt to the standard {@link SharedPreferences.OnSharedPreferenceChangeListener}
+ * @since 03.07.2019
```
Set directly to `@since 0.9.0`

davidgarciaanton commented:
Hi @patrickfav I've updated the PR with some of the corrections you requested, and added some Unit and Integration tests.

Let me know if I missed something.

patrickfav commented on armadillo/src/main/java/at/favre/lib/armadillo/SharedPreferenceChangeListenerWrapper.java:
```diff
@@ -43,14 +47,14 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
-        if (sharedPreferences != null && key != null) {
-            wrapped.onSecurePreferenceChanged(keyComparison, sharedPreferences, key);
+        if (key != null) {
+            wrapped.onSecurePreferenceChanged(securedPrefs, new KeyComparisonImpl(encryptionProtocol, key));
```
See comment above. I think it would be better to pass to:

    wrapped.onSecurePreferenceChanged(this, new KeyComparisonImpl(encryptionProtocol, key));

davidgarciaanton replied:
I'm not sure if I'm understanding correctly. I'm just passing the `securedPrefs` instance, with is actually initialised with the library `SecureSharedPreferences` instance, so that as you say in your comment the user doesn't have access to the raw encrypted data, but to the wrapper that will allow to get decrypted info.

patrickfav replied:
yes, I m not sure it is of much use to get the encrypted data, now would it?

davidgarciaanton replied:
Well, but actually the user will be able to access the decrypted data through armadillo's SecureSharedPreferences instance.

What is exactly the change you're proposing here?

patrickfav commented:
LGTM.
